### PR TITLE
Feature/add schema validation kv table sink

### PR DESCRIPTION
**Pull request by gsps1**
Added schema validation and tests in KVTable sink and also a minor bugfix in StreamBatchSource.

### CODE REVIEW

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 1]
```diff
@@ -80,8 +82,34 @@ public KVTableSink(KVTableConfig kvTableConfig) {
   }
 
   @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    // validate that input and output fields are present
+    SchemaValidator.validateFieldsArePresentInSchema(inputSchema, kvTableConfig.keyField);
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: only do this if the input schema is not null

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 2]
```diff
@@ -80,8 +82,34 @@ public KVTableSink(KVTableConfig kvTableConfig) {
...
+
+  private void validateSchemaTypeIsStringOrBytes(Schema inputSchema, String fieldName) {
+    Schema.Type fieldType = inputSchema.getField(fieldName).getSchema().getType();
+    if (!fieldType.equals(Schema.Type.STRING) && (!fieldType.equals(Schema.Type.BYTES))) {
+      throw new IllegalArgumentException(
+        String.format("Field name %s is of type %s, only types String and Bytes are supported for KVTable",
+                      fieldName, fieldType));
+    }
+  }
+  @Override
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java:113]: nit: add newline between methods

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 3]
```diff
@@ -80,8 +82,34 @@ public KVTableSink(KVTableConfig kvTableConfig) {
...
 
   @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    // validate that input and output fields are present
+    SchemaValidator.validateFieldsArePresentInSchema(inputSchema, kvTableConfig.keyField);
+    SchemaValidator.validateFieldsArePresentInSchema(inputSchema, kvTableConfig.valueField);
+    validateSchemaTypeIsStringOrBytes(inputSchema, kvTableConfig.keyField);
+    validateSchemaTypeIsStringOrBytes(inputSchema, kvTableConfig.valueField);
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: the value can be nullable but the key cannot.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 4]
```diff
@@ -91,9 +119,9 @@ public KVTableSink(KVTableConfig kvTableConfig) {
   public void transform(StructuredRecord input, Emitter<KeyValue<byte[], byte[]>> emitter) throws Exception {
     Object key = input.get(kvTableConfig.keyField);
     Preconditions.checkArgument(key != null, "Key cannot be null.");
-    byte[] keyBytes = key instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) key) : (byte[]) key;
+    byte[] keyBytes = key instanceof String ? Bytes.toBytes((String) key) : (byte[]) key;
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: can't get rid of the ByteBuffer logic. The previous logic should be applied if the schema says the field's type is bytes.  If the schema says the field's type is string, then you can cast it directly to a String and just do Bytes.toBytes() on it.

**Comment by gsps1** [comment]
@albertshau can you review again? addressed the comments. Thanks

**Comment by gsps1** [comment]
@albertshau have updated the PR. please review. Thanks

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 5]
```diff
@@ -91,9 +129,40 @@ public KVTableSink(KVTableConfig kvTableConfig) {
...
-    byte[] keyBytes = key instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) key) : (byte[]) key;
+    byte[] keyBytes;
+
+    Schema keyFieldSchema = input.getSchema().getField(kvTableConfig.keyField).getSchema();
+    if (keyFieldSchema.getType().equals(Schema.Type.STRING)) {
+      keyBytes = Bytes.toBytes((String) key);
+    } else if (keyFieldSchema.getType().equals(Schema.Type.BYTES)) {
+      keyBytes = key instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) key) : (byte[]) key;
+    } else if (keyFieldSchema.isNullable()) {
+      throw new Exception("Key is nullable in the input record : " + input);
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: don't log the record, as StructuredRecord has not overridden toString() as anything useful.  Message should be something like "Key field %s cannot have schema %s. It must be a bytes or string.".

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 6]
```diff
@@ -91,9 +129,40 @@ public KVTableSink(KVTableConfig kvTableConfig) {
...
+    } else {
+      throw new Exception(
+        "Key is of unsupported type in the record" + input + ", only String and Bytes are supported");
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: don't log the record, as StructuredRecord has not overridden toString() as anything useful. Message should be something like "Key field %s cannot have schema %s. It must be a bytes or string.".

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java (hunk 5 above)
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: "Key field %s cannot have nullable schema %s"

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 7]
```diff
@@ -91,9 +129,40 @@ public KVTableSink(KVTableConfig kvTableConfig) {
...
+    }
+
+    Schema.Field valueFieldSchema = input.getSchema().getField(kvTableConfig.valueField);
+    if (valueFieldSchema == null) {
+      throw new Exception("Field " + kvTableConfig.valueField + " is missing in the input record");
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: "Key field"

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java [hunk 8]
```diff
@@ -91,9 +129,40 @@ public KVTableSink(KVTableConfig kvTableConfig) {
...
+      Schema.Type valueFieldType =
+        valueFieldSchema.getSchema().isNullable() ? valueFieldSchema.getSchema().getNonNullable().getType() :
+          valueFieldSchema.getSchema().getType();
+      if (valueFieldType.equals(Schema.Type.STRING)) {
+        valBytes = Bytes.toBytes((String) val);
+      } else if (valueFieldType.equals(Schema.Type.BYTES)) {
+        valBytes = val instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) val) : (byte[]) val;
+      } else {
+        throw new Exception(
+          "Value is of unsupported type in the record, " + input + " only String and Bytes are supported");
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java]: "Value field %s has unsupported schema %s. Only String and Bytes are supported"

**Comment by albertshau** [comment]
remaining comments are only about exception messages, rest lgtm. Please fix then merge

**Pull request merged**